src/audio/base_fw.c

```diff
@@ -303,6 +317,38 @@ static int basefw_get_large_config(struct comp_dev *dev,
 	return -EINVAL;
 };
 
+#ifdef CONFIG_LOG_BACKEND_ADSP_MTRACE
+
+/**
+ * Maximum time to block IPC BUFFER STATUS notify messages.
+ * The notification is sent either due to amount of data
+ * available in the buffer, or by timeout in case logs are
+ * sent at slow rate.
+ */
+#define IPC4_MTRACE_NOTIFY_MAX_DELAY_MS 1000
+
+static uint64_t msg_notify_last_sent;
+
+static void mtrace_log_hook(size_t written, size_t space_left)
+{
+	if (space_left < NOTIFY_BUFFER_STATUS_THRESHOLD ||
+	    k_uptime_delta(&msg_notify_last_sent) >= IPC4_MTRACE_NOTIFY_MAX_DELAY_MS)
+		ipc_send_buffer_status_notify();
+
+	msg_notify_last_sent = k_uptime_get();
+}
+#endif
+
+static int mtrace_enable(void)
+{
+#ifdef CONFIG_LOG_BACKEND_ADSP_MTRACE
+	adsp_mtrace_log_init(mtrace_log_hook);
+	return 0;
+#else
+	return IPC4_UNKNOWN_MESSAGE_TYPE;
+#endif
+}
+
```

Comment:
Not entirely sure of the location. This is an interface of the basefw, but otherwise this code might be better in a separate file somewhere.

Reply:
I now moved this to a separate file as the amount of code kept increasing and it feels out of place in base_fw.c.